New RPA integral and memory savings

mkakcl opened the pull request:
Updated integral for the zeroth moment of RPA calculations.

Changes to the manner in which se_moments are build so that only the required information for the current dd-moment and for future moments is stored. This significantly reduces the memory overhead of mGW calculations.

mkakcl committed:
Updated integral for the zeroth moment of RPA calculations.

Changes to the manner in which se_moments are build so that only the required information for the current dd-moment and for future moments is stored. This significantly reduces the memory overhead of mGW calculations.

mkakcl committed:
Linting

mkakcl committed:
A bit more linting

mkakcl committed:
Should fix the errors in the tests.

Unrestricted is now using the new RPA integral, but not using the memory saving moment construction. Will investigate feasibility.

THC cannot use memory saving method, will need to keep its own kernel.

mkakcl committed:
Fixed a small bug related to uGW, see comment in PR for full details.

Will fix test regression values in next push based on CI

mkakcl committed:
Revert last commit, change the regression test values and added an RPA regression test for fsuGW to be in line with other self-consistent tests. See comment in PR for full details.

mkakcl committed:
Liniting

mkakcl committed:
Reimplement the previous uncommited changes, with some additional alterations to make it work as intended. Tests have been updated to reflect. See comment in PR.

mkakcl commented:
### dRPA uGW

Just wanted to explain what happened over the previous few commits and why the test regression values changed. 

A discrepancy was created between regression test values from the old integral code for RPA and the new code within uGW calculations. This reflected a bug in the original uGW code for RPA which had been included in the original commits for the new integral. Previously alpha and beta electrons were handled separately in the integral(s). However, in the calculation for Q and other terms, a sum over occupied-virtual orbitals occurs and thus both alpha and beta electrons need to be considered. This was somewhat accidentally corrected for by running the integrals with the restricted RPA code where an extra factor of 2 is included for spin. This caused the comparisons to the rGW code to be accurate and this generally only creates a small error if the alpha and beta states are similar, as was the case with the comparison to pyscf. The largest error would exist in the no beta test, however since we had nothing to compare to, the comparison is just to the code and thus was not noticed.

This should now be corrected in the latest PR with the integral being done for the concatenated terms and an extra spin option added into the integral to remove the factor of 2 when completing uGW calculations.

mkakcl committed:
Bring uGW in line with restricted in terms of memory saving.

mkakcl committed:
Linting.

basilib committed:
Merge branch 'master' into new_rpa

obackhouse commented on momentGW/pbc/uhf/tda.py:
```diff
@@ -33,6 +33,115 @@ class dTDA(KdTDA, MolUdTDA):
+
+    @logging.with_timer("Zeroth density-density moments")
+    @logging.with_status("Constructing zeroth density-density moment")
+    def build_zeroth_moment(self):
```
shouldn't these functions be called `build_zeroth_dd_moment`?

mkakcl replied:
Updated in latest commit

obackhouse commented on momentGW/pbc/uhf/tda.py:
```diff
@@ -209,23 +318,28 @@ def build_se_moments(self, moments_dd):
+                if moments_dd is None:
+                    recursion_term, eta_aux = self.build_nth_dd_moment(
+                        n, q, recursion_term, zeroth_mom
                     )
-                    eta_aux += np.dot(moments_dd[q, kb, n], Lia.T.conj())
+                else:
+                    eta_aux = 0
+                    for kj in kpts.loop(1, mpi=True):
+                        kb = kpts.member(kpts.wrap_around(kpts[q] + kpts[kj]))
+                        eta_aux += np.dot(moments_dd[q, kb, n], self.Lia(kj, kb).T.conj())
 
-                eta_aux = mpi_helper.allreduce(eta_aux)
-                eta_aux /= self.nkpts
+                    eta_aux = mpi_helper.allreduce(eta_aux)
+                    eta_aux /= self.nkpts
```
this is all tested with MPI? we don't do MPI tests on gh, maybe we should

mkakcl replied:
I quickly tested the molecular MPI not so much on the pbe and UGW. The next PR is going to change the way MPI is done for MPI in pbc, so with this, will add some tests to gh with MPI, since I agree would be good to have some.

obackhouse commented on momentGW/pbc/uhf/tda.py:
```diff
@@ -251,6 +365,26 @@ def build_se_moments(self, moments_dd):
+            ],
+            axis=1,
+        )
+
```
these are not standard interface functions right? not sure i'm a fan, if you really want them i'd make them private, maybe `_get_Lia` and `_get_Lai`

mkakcl replied:
Copilot PR

obackhouse commented on momentGW/rpa.py:
```diff
@@ -356,106 +316,64 @@ def eval_diag_main_integral(self, quad, d, diag_eri):
             The (aux, W occ, W vir) integral array. If `None`, use
             `self.integrals.Lia`. Keyword argument allows for the use of
             this function with `uhf` and `pbc` modules.
+        spin : bool
```
can you make the variable name more descriptive? `include_spin_factor` maybe?

mkakcl replied:
Copilot PR

obackhouse commented on momentGW/rpa.py:
```diff
@@ -465,32 +383,24 @@ def eval_main_integral(self, quad, d, Lia=None):
-            Quadrature points.
-        weights : numpy.ndarray
-            Quadrature weights.
+    def gen_ClenCur_quad_semiinf(self):
```
```suggestion
    def gen_clencur_quad_semiinf(self):
```
change throughout

mkakcl replied:
Copilot PR

obackhouse commented on momentGW/rpa.py:
```diff
@@ -465,32 +383,24 @@ def eval_main_integral(self, quad, d, Lia=None):
+                1.0 * (4 * np.sin(t) / ((self.gw.npoints + 1) * (1 - np.cos(t)) ** 2)) * s
+                for (t, s) in zip(tvals, jsums)
+            ]
+        )
```
these would be better if they were vectorised numpy operations rather than list comprehensions, if possible? maybe not too big a deal if it's not a bottleneck but i'd do it anyway

mkakcl replied:
Copilot PR

obackhouse commented on momentGW/thc.py:
```diff
@@ -266,6 +266,30 @@ class dTDA(DFdTDA):
+        moments_vir : numpy.ndarray
+            Moments of the virtual self-energy for each spin channel.
+        """
+        # Build the density-density response moments
+        moments_dd = self.build_dd_moments()
+
+        # Build the self-energy moments
+        moments_occ, moments_vir = self.build_se_moments(moments_dd)
+
+        return moments_occ, moments_vir
```
can you explain why this only looks different now for THC and not for the base class?

mkakcl replied:
It is a mix of there are some fundamental differences between THC and CD in terms of building moments, so there will be a difference. However in a broader sense, the THC code is getting a large restructuring with RPA and a better TDA code coming, so this is more of a temporary solution until the full refactor is introduced. I didn't want this PR to grow too large/have too much scope.

obackhouse commented on momentGW/uhf/rpa.py:
```diff
@@ -60,31 +61,19 @@ def integrate(self):
         diag_eri_β[b0:b1] = util.einsum("np,np->p", self.integrals[1].Lia, self.integrals[1].Lia)
         diag_eri_β = mpi_helper.allreduce(diag_eri_β)
-        diag_eri = (diag_eri_α, diag_eri_β)
+        diag_eri = np.concatenate([diag_eri_α, diag_eri_β])
```
is this necessary? it copies the arrays

mkakcl replied:
It is required in the context of needing to use the combined diag_eri to fit the integral since the integral needs to include Lia over both spins. Whether this is the best way to do this, see the comment below regards building the concatenated Lia "again, i'd avoid this at all costs -- ..."

obackhouse commented on momentGW/uhf/rpa.py:
```diff
@@ -60,31 +61,19 @@ def integrate(self):
-        quad = (
-            self.optimise_main_quad(d_full[0], diag_eri[0], name="Main (α)"),
-            self.optimise_main_quad(d_full[1], diag_eri[1], name="Main (β)"),
+        # Calculate (L|ia) D_{ia} and (L|ia) D_{ia}^{-1} intermediates
+        Lia = np.concatenate(
+            [
+                self.integrals[0].Lia,
+                self.integrals[1].Lia,
+            ],
+            axis=1,
         )
```
again, i'd avoid this at all costs -- i think that `np.concatenate` will copy the arrays and increase memory footprint

obackhouse replied:
i guess it may be tricky to avoid this... not a big deal for UHF where higher memory is kind of expected, so feel free to dismiss

mkakcl replied:
So everything you have said I do agree with and I think it is right. However, none of these changes are really new, (aside from the concatenated diag_eri but this was leading to an error in the code). If you look at `def build_dd_moments ` in uhf/rpa, which is the original way the moments were built, this is also done. It is also done in TDA but in a more intelligent way by just using the zeroth moment as Lia. So this isn't a new problem, it just appears more blatantly in this refactor. Though this refactor will save more than the factor of 2 that storing them twice like this does (assuming you run more than 2 moments).

More generally, I think what is really required is a refactoring of the uhf ints, such that at least Lia when built and called are just the concatenated integrals. Whenever we are using them, and doing V_{P,ia}V_{ia,Q}, which is how all moment build ends, we need to be summing over both. This lead to the error in the uGW RPA code in the integral, where they weren't summed over. In terms of the other integrals, Lpq and Lpx, I am less sure since I am generally a bit less informed on the UGW code.

obackhouse commented on examples/20-store_calculation.py:
```diff
@@ -0,0 +1,65 @@
+moments = rpa.build_se_moments(moments_dd)
+
+# Solve dyson's equation
+gw.kernel(nmom_max, integrals=integrals, moments=moments)
```
this example doesn't seem to store the calculation as it says it does?

mkakcl replied:
I think this should be a better example of how to store information from a calculation.

obackhouse replied:
but what i mean is that at no point is it actually demonstrating how to store+restore a calculation using this information

obackhouse replied:
anyway it's fine, but maybe we should have actual wrappers for writing the h5 file and reading it, i thought we had a base one for the solvers but i guess not

obackhouse reviewed (approved):
I've left comments from reading the code. As for the methodology and how it works, I trust your judgement there, I won't get time to properly dig into the changes until they're written up for a paper :smile:

obackhouse reviewed (changes requested):
.

obackhouse commented:
@copilot read over this PR and make any relevant suggestions based on: the consistency of the changes with respect to the existing style and choices in the codebase; any simple optimisation, particularly for numpy operations and array memory footprint; changes or additions related to the API and inheritance structure. Do not make any comment on the maths or methodology.

Copilot committed:
Initial plan

Copilot committed:
Improve code consistency and optimize numpy operations

- Rename gen_ClenCur_quad_semiinf to gen_clencur_quad_semiinf for consistency
- Rename 'spin' parameter to 'include_spin_factor' for clarity
- Add missing 'd' parameter to eval_main_integral docstring
- Vectorize quadrature generation to replace list comprehensions with numpy operations
- Make Lia and Lai helper methods private (_Lia, _Lai) in pbc/uhf/tda.py

Co-authored-by: obackhouse <34662120+obackhouse@users.noreply.github.com>

Copilot committed:
Remove excessive blank lines in gen_clencur_quad_semiinf

Co-authored-by: obackhouse <34662120+obackhouse@users.noreply.github.com>

Copilot committed:
Fix docstring inconsistencies in UHF modules

- Add missing Lia parameter documentation in build_nth_dd_moment()
- Fix parameter name from "zeroth moment" to "zeroth_mom" to match actual parameter
- Correct return type documentation: single numpy.ndarray instead of tuple
- Update integral parameter type to match actual implementation (single array, not tuple)

Co-authored-by: obackhouse <34662120+obackhouse@users.noreply.github.com>

Copilot committed:
Fix parameter name inconsistency in PBC modules

- Fix "zeroth moment" to "zeroth_mom" in docstrings for pbc/rpa.py, pbc/tda.py, and pbc/uhf/tda.py

Co-authored-by: obackhouse <34662120+obackhouse@users.noreply.github.com>

Copilot committed:
Add array shape documentation for Lia parameter

- Specify expected shape (naux, nocc_α * nvir_α + nocc_β * nvir_β) for Lia parameter in UHF modules

Co-authored-by: obackhouse <34662120+obackhouse@users.noreply.github.com>

mkakcl committed:
Linting

mkakcl committed:
Merge pull request #80 from BoothGroup/copilot/sub-pr-79

Improve code consistency, API clarity, and numpy performance

mkakcl committed:
Updates based on PR

mkakcl committed:
Update pyproject.toml

Put an upper limit on the version allowed for numpy since version 2.4.0 breaks the util.einsum function.

mkakcl committed:
Update pyproject.toml

Correction to version change
